File: src/main/java/com/warrantybee/api/services/implementations/CloudinaryStorageService.java
package com.warrantybee.api.services.implementations;

import com.cloudinary.Cloudinary;
import com.cloudinary.utils.ObjectUtils;
import com.warrantybee.api.services.interfaces.IStorageService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;

import java.util.Map;

/**
 * Implementation of {@link IStorageService} that uses Cloudinary
 * as the underlying storage provider.
 * <p>
 * This service handles file uploads and deletions by interacting
 * with Cloudinary's Java SDK. Uploaded files return a secure URL,
 * while deletions use the public ID provided during upload.
 * </p>
 */
@Service
public class CloudinaryStorageService implements IStorageService {

    private final Cloudinary _cloudinary;

    /**
     * Creates a new instance of {@code CloudinaryStorageService}.
     *
     * @param cloudinary the Cloudinary client instance injected by Spring
     */
    @Autowired
    public CloudinaryStorageService(Cloudinary cloudinary) {
        this._cloudinary = cloudinary;
    }

    /**
     * Uploads a file to Cloudinary and returns the secure URL of the uploaded file.
     *
     * @param file the multipart file to upload; must not be null
     * @return the secure URL of the uploaded file, or {@code null} if the upload fails
     */
    @Override
    public String upload(MultipartFile file) {
        try {
            Map<?, ?> options = ObjectUtils.emptyMap();
            Map<?, ?> uploadResult = _cloudinary.uploader().upload(file.getBytes(), options);
            return uploadResult.get("secure_url").toString();
        } catch (Exception e) {
            return null;
        }
    }

    /**
     * Deletes a file from Cloudinary using its public identifier.
     *
     * @param id the Cloudinary public ID of the file to delete
     * @return {@code true} if deletion is successful; {@code false} otherwise
     */
    @Override
    public Boolean delete(String id) {
        try {
            Map<?, ?> options = ObjectUtils.emptyMap();
            Map<?, ?> result = _cloudinary.uploader().destroy(id, options);
            return true;
        } catch (Exception e) {
            return false;
        }
    }

    /**
     * Extracts the Cloudinary {@code public_id} from a given Cloudinary resource URL.
     *
     * @param url the full Cloudinary file URL; may be null or empty
     * @return the extracted {@code public_id}, or {@code null} if the URL is invalid
     */
    public static String getPublicId(String url) {
        if (url == null || url.isEmpty()) {
            return null;
        }

        try {
            String[] parts = url.split("/upload/");
            if (parts.length < 2) {
                return null;
            }

            String path = parts[1];

            if (path.startsWith("v")) {
                path = path.substring(path.indexOf("/") + 1);
            }

            int dotIndex = path.lastIndexOf('.');
            if (dotIndex != -1) {
                path = path.substring(0, dotIndex);
            }

            return path;
        }
        catch (Exception e) {
            return null;
        }
    }
}

File: src/WarrantyBee.Infrastructure/Services/CloudinaryStorageService.cs
using CloudinaryDotNet;
using CloudinaryDotNet.Actions;
using Microsoft.Extensions.Options;
using WarrantyBee.Shared.Infrastructure.Abstractions;
using WarrantyBee.Shared.Core.Configuration;
using WarrantyBee.Application.Abstractions.Services;

namespace WarrantyBee.API.Infrastructure.Services;

/// <summary>
/// Provides storage services using Cloudinary as the backend.
/// </summary>
public class CloudinaryStorageService : IStorageService
{
    private readonly Cloudinary _cloudinary;

    /// <summary>
    /// Initializes a new instance of the <see cref="CloudinaryStorageService"/> class.
    /// </summary>
    /// <param name="config">The application configuration containing Cloudinary settings.</param>
    /// <exception cref="ArgumentNullException">Thrown when configuration or Cloudinary settings are missing.</exception>
    public CloudinaryStorageService(IOptions<AppConfiguration> config)
    {
        var cfg = config.Value.Cloudinary ?? throw new ArgumentNullException(nameof(config));
        var account = new Account(cfg.Cloud, cfg.ApiKey, cfg.ApiSecret);
        _cloudinary = new Cloudinary(account);
    }

    /// <summary>
    /// Uploads a file to Cloudinary asynchronously.
    /// </summary>
    /// <param name="fileStream">The stream of the file to upload.</param>
    /// <param name="fileName">The name of the file.</param>
    /// <param name="contentType">The content type of the file.</param>
    /// <returns>The secure URL of the uploaded file, or an empty string if the upload fails.</returns>
    public async Task<string> UploadAsync(Stream fileStream, string fileName, string contentType)
    {
        var uploadParams = new ImageUploadParams
        {
            File = new FileDescription(fileName, fileStream),
            UseFilename = true,
            UniqueFilename = true,
            Overwrite = false
        };

        var uploadResult = await _cloudinary.UploadAsync(uploadParams);
        return uploadResult.SecureUrl?.ToString() ?? string.Empty;
    }

    /// <summary>
    /// Deletes a file from Cloudinary by its public identifier.
    /// </summary>
    /// <param name="id">The public identifier of the file to delete.</param>
    /// <returns><c>true</c> if the file was successfully deleted; otherwise, <c>false</c>.</returns>
    public async Task<bool> DeleteAsync(string id)
    {
        var deletionParams = new DeletionParams(id);
        var result = await _cloudinary.DestroyAsync(deletionParams);
        return result.Result == "ok";
    }

    /// <summary>
    /// Deletes a file from Cloudinary by its URL.
    /// </summary>
    /// <param name="url">The secure URL of the file to delete.</param>
    /// <returns><c>true</c> if the file was successfully deleted; otherwise, <c>false</c>.</returns>
    public async Task<bool> DeleteByUrlAsync(string url)
    {
        var publicId = GetPublicId(url);
        if (string.IsNullOrWhiteSpace(publicId)) return false;
        return await DeleteAsync(publicId);
    }

    /// <summary>
    /// Extracts the public identifier of a file from its Cloudinary URL.
    /// </summary>
    /// <param name="url">The Cloudinary URL.</param>
    /// <returns>The public identifier if found; otherwise, <c>null</c>.</returns>
    private static string? GetPublicId(string? url)
    {
        if (string.IsNullOrWhiteSpace(url)) return null;

        try
        {
            var parts = url.Split("/upload/");
            if (parts.Length < 2) return null;

            var path = parts[1];
            if (path.StartsWith("v"))
            {
                var slashIndex = path.IndexOf("/");
                if (slashIndex != -1) path = path.Substring(slashIndex + 1);
            }

            var dotIndex = path.LastIndexOf('.');
            if (dotIndex != -1) path = path.Substring(0, dotIndex);

            return path;
        }
        catch
        {
            return null;
        }
    }
}
